### ISSUE TITLE
Bundle() inline verification of Intoto Statements

### ISSUE THREAD

**Issue by ramonpetgrave64** [sigstore/sigstore-go#256]
I'm getting started with using the `Bundle()` function to make a signed Sigstore Bundle from an intotoStatement.

I'm invoking similar to 

```
statementBytes, _ := json.Marshal(*statement)
content := &sigstoreSign.DSSEData{
    Data:        statementBytes,
    PayloadType: "application/vnd.in-toto+json",
}
bundleOpts := &sigstoreSign.BundleOptions{
    ...
    TruestedRoot: trustedRoot,
}
bundle, _ := sigstoreSign.Bundle(content, keypair, *bundleOpts)
...
```

If I specify a `TrustedRoot`, then `Bundle() `will also [attempt](https://github.com/sigstore/sigstore-go/blob/6f7e99d5908575d5faaddb5a688f83430ba989b1/pkg/sign/signer.go#L141) to `Verify()` before returning the bundle. But I think there may be a few problems with the setup for there.

1. The [options](https://github.com/sigstore/sigstore-go/blob/6f7e99d5908575d5faaddb5a688f83430ba989b1/pkg/sign/signer.go#L152) `artifactOpts := verify.WithArtifact(bytes.NewReader(content.PreAuthEncoding()))` may not be correct to use, since it treats the "artifact" as the envelope, not as the artifact(s) that the envelope references.
1. `verify.WithArtifactDigest()` is probably more appropriate, but it does not currently support [multiple artifacts](https://github.com/sigstore/sigstore-go/blob/6f7e99d5908575d5faaddb5a688f83430ba989b1/pkg/verify/signed_entity.go#L456-L459).
1. I'm a still new to go, so I'm not sure about this one: Since the envelope's Digest (<algo>: <hash>)will be a`map[string]string`, and not `map[string]byte` does the hash need to be [hex-decoded ](https://github.com/sigstore/sigstore-go/blob/6f7e99d5908575d5faaddb5a688f83430ba989b1/pkg/verify/signature.go#L209)into bytes before comparison?

**Comment by ramonpetgrave64** [comment]
@codysoyland

**Comment by codysoyland** [comment]
> 1. The [options](https://github.com/sigstore/sigstore-go/blob/6f7e99d5908575d5faaddb5a688f83430ba989b1/pkg/sign/signer.go#L152) `artifactOpts := verify.WithArtifact(bytes.NewReader(content.PreAuthEncoding()))` may not be correct to use, since it treats the "artifact" as the envelope, not as the artifact(s) that the envelope references.

That's right, this looks like a bug to me. I believe this is the correct code if you supply `PlainData` content, but `WithArtifact`/`WithArtifactDigest` checks against the subject digest in the intoto statement.

> 3. `verify.WithArtifactDigest()` is probably more appropriate, but it does not currently support [multiple artifacts](https://github.com/sigstore/sigstore-go/blob/6f7e99d5908575d5faaddb5a688f83430ba989b1/pkg/verify/signed_entity.go#L456-L459).

Agreed. Right now, we don't support multiple subjects for a given attestation bundle.

> 5. I'm a still new to go, so I'm not sure about this one: Since the envelope's Digest (: )will be a`map[string]string`, and not `map[string]byte` does the hash need to be [hex-decoded ](https://github.com/sigstore/sigstore-go/blob/6f7e99d5908575d5faaddb5a688f83430ba989b1/pkg/verify/signature.go#L209)into bytes before comparison?

Yes, the [digest fields are hex encoded](https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md), and therefore must be decoded to bytes before comparison, like the code you linked.

I filed https://github.com/sigstore/sigstore-go/issues/257 to capture the specific bug here.

**Comment by ramonpetgrave64** [comment]
Thanks for following up

1. It looks like if I use `PlainData`, the bundle won't embed an `Envelope`, which I need for my use-case.

**Comment by Hayden-IO** [comment]
Should we file an issue for a “WithArtifacts/WithArtifactDigests” to handle multiple subjects?

**Comment by steiza** [comment]
Oops! Sorry about that @ramonpetgrave64, this was definitely a mistake on my part. I believe this should be resolved with #258. I'm trying to avoid having sigstore-go signing be aware of what DSSE contains, and so that change will work with an in-toto attestation with multiple subjects.

In terms of verifying generally with multiple artifacts, one option would be to run verification with each artifact / artifact digest you care about. It certainly would be possible for sigstore-go verification to accept multiple artifacts / artifact digests, but if we're in a situation where some artifacts succeeded and others failed we'd have to think carefully about how to convey that to the user.

**Comment by ramonpetgrave64** [comment]
Thanks for fixing this so quickly.
